### PR TITLE
WebAuthn WD-07 api tests

### PR DESCRIPTION
**Pull request by apowers313**
new tests based on Working Draft 7

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/23445)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/334896425)

File: webauthn/securecontext.http.html [hunk 1]
```diff
@@ -17,7 +17,7 @@
     // Example 1
     // http://example.com/ opened in a top-level browsing context is not a secure context, as it was not delivered over an authenticated and encrypted channel.
     test (() => {
-        assert_false (typeof navigator.credentials.create === "function");
+        assert_false (typeof navigator.credentials === "object" && typeof navigator.credentials.create === "function");
```
Comment (jcjones) [inline webauthn/securecontext.http.html]: Note: This throws on line helpers.js:426 in `ensureInterface` because `navigator.credentials` doesn't exist, so `navigator.credentials.create` throws with an uncaught TypeError.

Perhaps that could be a separate commit?

Comment (apowers313) [reply webauthn/securecontext.http.html]: I had fixed i in one file but not the other... should be right in a0a0e8c

Comment (jcjones) [reply webauthn/securecontext.http.html]: Sorry, the issue is in `helpers.js` line 450 (now):

https://github.com/w3c/web-platform-tests/blob/4d8efcb6b6f4b784989c23cf83f46255388871a3/webauthn/helpers.js#L372-L374

File: webauthn/createcredential-badargs-user.https.html [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    new CreateCredentialsTest("options.publicKey.user.id", {}).runTest("Bad user: id is object", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", null).runTest("Bad user: id is null", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", "").runTest("Bad user: id is empty String", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", new Array()).runTest("Bad user: id is empty Array", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", new ArrayBuffer(0)).runTest("Bad user: id is empty ArrayBuffer", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", new ArrayBuffer(65)).runTest("Bad user: id is too long (65 bytes)", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", new Int16Array(33)).runTest("Bad user: id is too long (66 bytes)", new TypeError());
+    new CreateCredentialsTest("options.publicKey.user.id", new Int32Array(17)).runTest("Bad user: id is too long (68 bytes)", new TypeError());
```
Comment (jcjones) [inline webauthn/createcredential-badargs-user.https.html]: These duplicate "Bad user: id is too long (68 bytes)", "Bad user: id is too long (65 bytes)" tests make for test harness errors on Firefox; they'll need de-duped names.

File: webauthn/createcredential-timeout.https.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+<script src="/resources/testharnessreport.js"></script>
+<script src=helpers.js></script>
+<body></body>
+<script>
+standardSetup(function() {
+    "use strict";
+
+    // bad timeout values
+    new CreateCredentialsTest({path: "options.publicKey.timeout", value: -1}).runTest("Bad timeout: negative", new TypeError());
+    new CreateCredentialsTest({path: "options.publicKey.timeout", value: 4294967295 + 1}).runTest("Bad timeout: too big", new TypeError());
```
Comment (jcjones) [inline webauthn/createcredential-timeout.https.html]: The spec doesn't impose an upper limit on the timeout size; I don't know if this is a valid test to assume it should be a TypeError. (Step 4 of [Section 5.1.3](https://w3c.github.io/webauthn/#createCredential))

**Review by jcjones — COMMENTED** [review]
This looks pretty much good to go. I've got those two comments, and otherwise I think we're good to merge. Thanks, Adam!

File: webauthn/createcredential-badargs-user.https.html (hunk 2 above)
Comment (apowers313) [reply webauthn/createcredential-badargs-user.https.html]: Good catch, thanks.

File: webauthn/createcredential-timeout.https.html (hunk 3 above)
Comment (apowers313) [reply webauthn/createcredential-timeout.https.html]: The `timeout` property is defined as `unsigned long`. The value is MAX_UNSIGNED_LONG + 1. I'll comment it out with a TODO while I research the proper behavior. :)

Comment (jcjones) [reply webauthn/createcredential-timeout.https.html]: Sounds good. That one might be tricky!

Comment (apowers313) [reply webauthn/createcredential-timeout.https.html]: fixed in 60f7ad1

File: webauthn/createcredential-badargs-user.https.html (hunk 2 above)
Comment (apowers313) [reply webauthn/createcredential-badargs-user.https.html]: fixed in e5da0e2

File: webauthn/helpers.js [hunk 4]
```diff
@@ -10,6 +9,160 @@ const cose_crv_P256 = 1;
...
+
+function createCredential(opts) {
+    opts = opts || {};
+
+    // set the default options
+    var createArgs = cloneObject(createCredentialDefaultArgs);
+    let challengeBytes = new Uint8Array(16);
+    window.crypto.getRandomValues(challengeBytes);
+    createArgs.options.publicKey.challenge = challengeBytes;
+    createArgs.options.publicKey.user.id = new Uint8Array();
```
Comment (jcjones) [inline webauthn/helpers.js]: This has to be initialized, or the test `Bad user: id is empty ArrayBuffer` is a lie 😁 
So let's make this `new Uint8Array(16);`

Comment (jcjones) [reply webauthn/helpers.js]: Actually, we need to do the same thing in CreateCredentialsTest after the line:
`         // cloneObject can't clone the BufferSource in user.id, so let's recreate it.`

**Review by jcjones — CHANGES_REQUESTED** [review]

File: webauthn/createcredential-excludecredentials.https.html [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src=helpers.js></script>
+<body></body>
+<script>
+standardSetup(function() {
+    "use strict";
+
+    // bad excludeCredentials values
+    new CreateCredentialsTest({path: "options.publicKey. excludeCredentials", value: undefined}).runTest("Bad user:  excludeCredentials missing", new TypeError());
```
Comment (jcjones) [inline webauthn/createcredential-excludecredentials.https.html]: Unexpected spaces in these `path`s

Comment (jcjones) [reply webauthn/createcredential-excludecredentials.https.html]: Also: I think setting `excludeCredentials` to `undefined` is valid, actually. That's as if we didn't pass the array, which is legal. So I think line 15 goes away, and 16 + 17 have that space trimmed.

File: webauthn/createcredential-timeout.https.html [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src=helpers.js></script>
+<body></body>
+<script>
+standardSetup(function() {
+    "use strict";
+
+    // bad timeout values
+    new CreateCredentialsTest({path: "options.publicKey.timeout", value: -1}).runTest("Bad timeout: negative", new TypeError());
```
Comment (jcjones) [inline webauthn/createcredential-timeout.https.html]: I don't think we can test this, either, actually. In Gecko at least, since the timeout is a unsigned integer, if JS gives a signed integer it gets converted to unsigned (lossily) and arrives into at least my WebAuthn code as valid.

I think if we wanted to check this, we'd have to define the IDL to be signed.

File: webauthn/getcredential-timeout.https.html [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+<script src=helpers.js></script>
+<body></body>
+<script>
+standardSetup(function() {
+    "use strict";
+
+    var credPromise = createCredential();
+
+    // bad timeout values
+    new GetCredentialsTest({path: "options.publicKey.timeout", value: -1})
```
Comment (jcjones) [inline webauthn/getcredential-timeout.https.html]: As before; this might not be testable.

File: webauthn/createcredential-excludecredentials.https.html (hunk 5 above)
Comment (apowers313) [reply webauthn/createcredential-excludecredentials.https.html]: fixed in 9a708e1

File: webauthn/helpers.js (hunk 4 above)
Comment (apowers313) [reply webauthn/helpers.js]: fixed in 87ed972

**Review by jcjones — APPROVED** [review]

**Comment by jcjones** [comment]
For me there are still a bunch of type coercion issues, but this is a great place to merge and make sure everyone's on the same page. Thanks so much, Adam!

**Comment by apowers313** [comment]
Yea, totally agree -- this is a good first guess and we can iterate from here. I'm also wondering if people handle errors the same way. I'm thinking we will find out pretty quickly.

**Comment by jcjones** [comment]
Oops, one more change needed to `helpers.js`:
Line 65 in createCredential:
```
    createArgs.options.publicKey.user.id = new Uint8Array();
```
needs to also be non-zero. Something like 
```
    createArgs.options.publicKey.user.id = new Uint8Array(2);
```

**Comment by apowers313** [comment]
I had made the change but didn't save the file... it's in cdbe7ce now.
